std/digest/package.d

```diff
@@ -1123,12 +1123,12 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
     auto secret = "A7GZIP6TAQA6OHM7KZ42KB9303CEY0MOV5DD6NTV".representation;
     auto data = "data".representation;
 
-    string hex1 = data.hmac!SHA1(secret).toHexString;
-    string hex2 = data.hmac!SHA1(secret).toHexString;
-    string hex3 = "data1".representation.hmac!SHA1(secret).toHexString;
+    auto hex1 = data.hmac!SHA1(secret).toHexString;
```

Comment:
This was a documented example. You just broke a public API however bad it might have been.

This needs a *safe* fix that *doesn't break* API.

Reply:
Please weigh in on https://issues.dlang.org/show_bug.cgi?id=12625

Reply:
I'm fine with language change, there is no easy way to avoid breakage in all possible cases.
We should still try to find a way to fix library code so that public examples still work.
In other words toHexString may need to dup internal static array iff assigned to string, some wrapper struct should be able to do that.

Reply:
I'm ok with both the language change and this fix. I don't think we can move forward without breaking code, because this wrapper struct would need to behave both like a `string` and like a `char[2*N]`.

The only possibility for a less invasive breaking change is to change the language like so:
```D
char[16] get();

// This gets lowered like so:
string s = get();

// Ok, value types without indirections implicitly convert to `const` or `immutable`
immutable(char[16]) __tmp = get();

// Ok slicing local static array variables is allowed and the result is inferred as `scope`.
scope immutable(char)[] s = __tmp[];
```

The above is similar to the C++ rule for extending rvalue lifetime:
```C++
string getString();
const string& s = getString(); // the rvalue lifetime gets extended to the lifetime of `s`
```

https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/

@WalterBright what do you think?

Reply:
> This was a documented example. You just broke a public API however bad it might have been.

In many cases such usage of `toHexString` has simply been wrong (we saw some instances of this problem on bugzilla, D.learn and even in the oauth code.dlang.org package). Also when I originally developed this `char[2n] toHexString` API it was only meant as a non-allocating alternative to the string version. The string overload should have been preferred whenever a string is needed, but as I wasn't aware of the implicit slicing feature this API design mistake slipped in.

I think with the new scope semantics ```string s = fuctionReturnsCharN();``` is simply incorrect (and has always been unsafe) and ```scope string s = fuctionReturnsCharN();``` is the correct way to write this. https://github.com/dlang/phobos/pull/5714 marks the old behaviour as deprecated and I think this is a reasonable solution.

(Only remaining question is: This is a `@system` function. Is there / should there be any difference regarding `scope` rules for `system/safe` functions?)


> This needs a safe fix that doesn't break API.

If possible, such a better solution should be used. But as @ZombineDev already mentioned I don't think we can make a wrapper which works as a replacement for `char[n]` and `string` at the same time? Also if your wrapper always calls `.idup` when generating a string, it may allocate memory even though a `scope string` would have been perfectly acceptable. I prefer a deprecation message to make users aware of this choice and let them choose the correct approach on a case-by-case basis.

Reply:
Okay, probably we can break it it to avoid twarting development of scope feature in compiler on the unceratin condition of if we can find a better fix.

Reply:
There is no way to tell it's being assigned to a string, as the return type doesn't play into the overload set.

As with other moves in phobos, I would say explicit duplication looks better than hiding the fact that the type is not what you think it is using auto.

So it might be better as:

```D
string hex2 = data.hmac!SHA1(secret).toHexString.idup;
```

Reply:
@schveiguy Actually, there is a way (sort of), but it involves DIP66 - multiple alias this:
```D
struct Result(size_t N)
{
    private char[2*N] _data;
    string getString() { return _data.idup; }
    ref char[2*N] getStaticArray() scope return { return _data; }

    // Even without DIP66, I hope we can get the `alias this : member`
    // syntax sooner, rather than later.
    alias this : getString, getStaticArray;
}
``` 

Anyway, I think extending the lifetime approach is much cleaner and future looking (in the context of the language as a whole).

Reply:
Yes, I think your first option is much cleaner.

Reply:
> You just broke a public API

Are you sure? I don't see how the API is broken. The call is completely possible even with the change to the compiler. What is broken is the implicit slicing of the rvalue result (which is, of course, invalid).

In fact, I wouldn't be surprised at all if in the prior version of the code, `hex1`, `hex2`, and `hex3` all point at the same data.